File: python/unitedColor/rgbColorToName.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-



def rgbColorToNameTranslate(rgbColor):
	# cette fonction retourne un tableau de noms
	
	# le fichier de correspondance des couleurs est ICI
	colorRefPath = '../../'
	colorRefName = 'color_rgb_table'
	colorRefFile = colorRefPath + colorRefName	
	
	colorAlias = []
	# lecture du fichier
	with open(colorRefFile, mode='rt') as f:
		# pour chaque ligne…
		for l in f:
			# lecture du code de la couleur, en fin de ligne
			# attention, le dernier caractère est un "saut de ligne"
			code = l[len(l) - 7:len(l) - 1]
			color = ''
			# si le code correspond à la requête…
			if code == rgbColor:
				# lecture du nom de la couleur correspondante
				end = False
				for c in l:
					if c != ':' and not end:
						color = color + c
					else:
						end = True
			# ajout de la couleur au tableau
			if len(color) != 0:
				colorAlias.append(color)
	return colorAlias



def verifyCode(rgbColor):
	# vérification de la validité de la saisie au clavier
	
	# liste des caractères autorisés pour écrire un nombre héxadécimal
	hexaDigit ='0123456789abcdef'
	# si la longueur n'est pas 6 : erreur de saisie…
	if len(rgbColor) != 6:
		print('An hexadecimal RGB color must have 6 digits!')
		return False
	# si un caractère est hors de la liste autorisée : erreur de saisie…
	for d in rgbColor:
		if not (d in hexaDigit):
			print("Authoryzed digits are '0' → '9' and 'a' → 'f'.")
			print("('A' → 'F' can be used, too.)")
			return False
	# pas d'erreur…
	return True



def readColorCode():
	# saisie au clavier du "code couleur" héxadécimal
	
	foo = False
	# tant que 'foo' est faux:
	while not foo:
		rgbColor = input('Please, enter an hexadecimal RGB color: ')
		# transformation des majuscules en minuscules
		rgbColor = rgbColor.lower()
		# vérification de la saisie : 6 "digits" hégadécimaux…
		foo = verifyCode(rgbColor)
	# 'foo' est vrai :
	# fin de la boucle
	return rgbColor



# main function
def rgbColorToName():
	myColorCode = readColorCode()
	myColorName = rgbColorToNameTranslate(myColorCode)
	print(myColorName)
	n = len(myColorName)
	if n < 1:
		msg = 'The "' + myColorCode + '" color is unnamed.'
		print(msg)
	else:
		msg = 'The "' + myColorCode + '" color is called "'
		for i in range(1, n + 1):
			print (i)
			if i == n:
				msg = msg + myColorName[i - 1] + '".'
			elif i == n - 1:
				msg = msg + myColorName[i - 1] + '" or "'
			else:
				msg = msg + myColorName[i - 1] + '", "'
		print(msg)



if __name__ == '__main__':
	rgbColorToName()

File: ruby/unitedColor/rgbColorToName.rb
#!/usr/bin/env ruby -w
# -*- coding: utf-8 -*-



def rgbColorToNameTranslate(rgbColor)
	# cette méthode retourne un tableau de noms
	
	# le fichier de correspondance des noms est ICI
	colorRefPath = '../../'
	colorRefName = 'color_rgb_table'
	colorRefFile = colorRefPath + colorRefName
	
	colorAlias = []
	# lecture du fichier
	f = File.new(colorRefFile, 'rt')
	l = f.readlines()
	n = l.length
	for i in (0..n-1) do
		# lecture du code de la couleur, en fin de ligne
		# attention, le dernier caractère est un "saut de ligne"
		j = l[i].length
		code = l[i][j-7..j-2]
		color = ''
		# si le code correspond à la requête…
		if code == rgbColor
			# lecture du nom de la couleur correspondante
			foo = false
			for k in (0..j-8) do
				c = l[i][k]
				if c != ':' and not foo
					color += c
				else
					foo = true
				end
			end
		end
		# ajout de la couleur au tableau
		if color.length != 0
			colorAlias.push(color)
		end
	end
	return colorAlias
end



def verifyCode(rgbColor)
	# liste des caractères autorisés pour écrire un nombre héxadécimal
	hexaDigit = '0123456789abcdef'
	# si la longueur n'est pas 6 : erreur de saisie…
	if rgbColor.length() != 6
		puts('An hexadecimal RGB color must have 6 digits!')
		# erreur :
		return true
	end
	# si un caractère est hors de la liste autorisée : erreur de saisie…
	rgbColor.each_char do |d|
		if hexaDigit.match?(d) == false
			puts('Authorized digits are \'0\’ → \'9\' and \'a\' → \'f\'.')
			puts('(\'A\' → \'F\' can be used, too.)')
			# erreur :
			return true
		end
	end
	# pas d'erreur :
	return false
end



def readColorCode()
	# saisie au clavier du "code couleur" héxadécimal
	
	err = true
	while err
		puts('Please, enter an hexadecimal RGB color:')
		rgbColor = gets.chomp
		# transformation des majuscules en minuscules
		rgbColor = rgbColor.downcase()
		# vérification de la saisie : 6 "digits"
		err = verifyCode(rgbColor)
	end
	# pas d'erreur
	# fin de la boucle
	return rgbColor
end



# main method • méthode principale
def rgbColorToName()
	myColorCode = readColorCode()
	myColorName = rgbColorToNameTranslate(myColorCode)
	n = myColorName.length()
	if n < 1
		msg = 'The "' + myColorCode + '" color is unnamed.'
		puts(msg)
	else
		msg = 'The "' + myColorCode + '" color is called "'
		(1..n).each do |i|
			if i == n
				msg = msg + myColorName[i - 1] + '".'
			elsif i == n - 1
				msg = msg + myColorName[i - 1] + '" or "'
			else
				msg = msg + myColorName[i - 1] + '", "'
			end
		end
		puts(msg)
	end
end


if __FILE__ == $0
	rgbColorToName()
end
